### DISCUSSION TITLE
cascade attempting to insert when it shouldn't

### DISCUSSION THREAD

**Discussion by milesje** [mikro-orm/mikro-orm#3114] [Q&A]
When performing `em.persistAndFlush(entity)` it is attempting to do an insert for a ManyToOne relation, but it fails because it is already in the database.

**Stack trace**
```
ERROR [ExceptionsHandler] insert into "cust-test"."vehicle_type" ("description", "id", "unit_type") values ('Tractor Conv (TA)', '318f4a49-135f-4d54-967a-2e2772c12d37', 'PU') returning "id", "description", "unit_type" - duplicate key value violates unique constraint "vehicle_type_pkey"
UniqueConstraintViolationException: insert into "cust-test"."vehicle_type" ("description", "id", "unit_type") values ('Tractor Conv (TA)', '318f4a49-135f-4d54-967a-2e2772c12d37', 'PU') returning "id", "description", "unit_type" - duplicate key value violates unique constraint "vehicle_type_pkey"
    at PostgreSqlExceptionConverter.convertException (/home/jmiles/code/market-edge/market-edge-ws-nestjs/node_modules/@mikro-orm/postgresql/PostgreSqlExceptionConverter.js:28:24)
    at PostgreSqlDriver.convertException (/home/jmiles/code/market-edge/market-edge-ws-nestjs/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:173:54)
    at /home/jmiles/code/market-edge/market-edge-ws-nestjs/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:177:24
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at PostgreSqlDriver.nativeInsert (/home/jmiles/code/market-edge/market-edge-ws-nestjs/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:161:21)
    at ChangeSetPersister.persistNewEntity (/home/jmiles/code/market-edge/market-edge-ws-nestjs/node_modules/@mikro-orm/core/unit-of-work/ChangeSetPersister.js:85:21)
    at ChangeSetPersister.executeInserts (/home/jmiles/code/market-edge/market-edge-ws-nestjs/node_modules/@mikro-orm/core/unit-of-work/ChangeSetPersister.js:29:13)
    at ChangeSetPersister.runForEachSchema (/home/jmiles/code/market-edge/market-edge-ws-nestjs/node_modules/@mikro-orm/core/unit-of-work/ChangeSetPersister.js:68:13)
    at UnitOfWork.commitCreateChangeSets (/home/jmiles/code/market-edge/market-edge-ws-nestjs/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js:657:9)
    at UnitOfWork.persistToDatabase (/home/jmiles/code/market-edge/market-edge-ws-nestjs/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js:609:13)
```

**Expected behavior**
I would have expected the relational data to be either skipped or an update to have been performed.

**Additional context**
This is an NestJS project where I have a database table that it is  already populated with vehicle types. I'm attempting to insert a new Vehicle with the vehicle type from the populated data associated with it.

```
const entity = await this.em.create(Vehicle, data, {schema});
entity.vehicleType = await this.em.getReference(VehicleType, vehicleTypeId);
await this.em.persistAndFlush(entity);
```
This code is attempting to perform an insert on the VehicleType (vehicle_type) table instead of just associating the referenced entity.

**Versions**

| Dependency | Version |
| - | - |
| node | 18.1.0 |
| typescript | 4.6.4 |
| mikro-orm | 5.1 .3 |
| pg | 8.7.3 |
| knex | 2.0.0 |

**Comment by milesje** [comment]
Here is a shortend version of the 2 entities in question....
Vehicle
```
@Entity({schema: "*"})
export class Vehicle {
  @PrimaryKey({ type: 'uuid', defaultRaw: 'gen_random_uuid()' })
  id: string
....
  @ManyToOne(() => VehicleType, {cascade: []})
  vehicleType: VehicleType
```

VehicleType
```
@Entity({schema: "*"})
export class VehicleType {
  @PrimaryKey({ type: 'uuid', defaultRaw: 'gen_random_uuid()' })
  id: string
  @Property({ default: '' })
  description: string
  @Property({ default: '' })
  unitType: string

  @OneToMany(() => Vehicle, (vehicles) => vehicles.vehicleType)
  vehicles = new Collection<Vehicle>(this)
}
```

**Reply by milesje** [reply]
When I attempt to "update" an entity it is also performing an insert instead of an update.... I'm assuming that the persistAndFlush() function would perform the correct database function insert/update depending on if the entity is already stored in the database but this does not appear to be what is happening. It is ALWAYS performing an insert.

**Reply by milesje** [reply]
I feel like I'm missing something quite vital here....

**Comment by B4nan** [comment]
The code you provided on its own will never end up with an insert statement. Please provide full reproduction so its clear what exactly are you doing, by doing that you will probably get to the culprit yourself - but it is in your code, not in the ORM. Apparently the `VehicleType` entity is not managed, as only not managed entities will produce insert queries. Only managed entities can produce update queries - you need to load them and change their state, that's how the UoW works.

My guess is that your problem is already in the `em.create()` call, where you probably pass the `vehicleType` property as object - it does not matter if there is a PK or not, this will make the relation to cascade persist and produce an insert query. Either remove that property from your schema or change it to be a naked PK instead of object:

```ts
// this is what you are probably doing, and it will indeed insert the relation entity too
const ent = em.create(Ent, { relation: { id: 1, name: '123' } });
await em.persist(ent).flush();

// while this won't
const ent = em.create(Ent, { relation: 1 });
await em.persist(ent).flush();

// neither this
const data = { relation: { id: 1, name: '123' } });
const relId= data.relation.id;
delete data.relation;
const ent = em.create(Ent, data);
ent.relation = em.getReference(Rel, relId);
await em.persist(ent).flush();
```

Note that neither `em.create()` nor `em.getReference` are async - there is no point awaiting them.

**Reply by milesje** [reply]
The vehicle type is NOT contain the vehicleType in the create statement which is why I'm calling the `entity.vehicleType = await this.em.getReference(VehicleType, vehicleTypeId)`

The full code of the function is 
```
 @UseRequestContext()
  async save(vehicleTypeId: string, vehicle: VehicleDto, schema: string) {
    const entity = await this.em.create(Vehicle, vehicle, { schema })
    entity.vehicleType = await this.em.getReference(VehicleType, vehicleTypeId)
    
    await this.em.persistAndFlush(saved)
    return saved
  }
```

I assume that the em.getReference() function would load the correct VehicleType from the db making it a managed entity.
If I print out the entity prior to the persistAndFlush call it does show a `vehicleType: Ref<VehicleType> { id: '71c6b3c6-75ab-43d2-be5a-9ec992b4143e' }`

**Reply by milesje** [reply]
This would a closer example of what I'm doing.... the data being passed into the em.create does NOT contain the vehicleType. The vehicleType is ONLY being added to the entity using the em.getReference() call.
```
const ent = await em.create(Ent, {make: 'Ford", model: 'F150', vinNumber: 'TX12763225615`, active: true}, {schema})
ent.vehicleType = await em.getReference(VehicleType, vehicleTypeId)
await em.presist(ent).flush()
```

**Reply by milesje** [reply]
here is the output of the entity just prior to the persistAndFlush() call.
```
Vehicle {
  locNum: '0',
  owner: 'O',
  unitNum: '1',
  year: '2022',
  make: 'Ford',
  model: 'F-150',
  vinNumber: 'TX1233563566',
  vehicleType: Ref<VehicleType> { id: '01fc4b1e-82e4-471a-a0ac-5d0001a85f81' }
}
```
I would assume that at this point the vehicleType is fully managed as it does contain the reference to the correct object. But it is attempting to perform an insert on the vehicle_type table which fails due to the fact that there is already an entry in the table with that PK.

**Reply by B4nan** [reply]
Still what I said applies, this code on its own will not do what you say. Try to create isolated reproduction and you will see yourself. 

You might as well have some problem with contexts, seeing you have the UseRQ decorator. It should be used only at the top level, not in service level, definitely not in methods you call from other decorated methods.

**Reply by milesje** [reply]
I'm not sure how to isolate it any more than it already is... it is a single function... when I print the entity object it appears correct to me.. it does not contain any extra data and DOES contain the reference to the vehicleType.

**Reply by B4nan** [reply]
Isolate it outside of your project. Simplify till it start to work. Create a reproduction. Otherwise I can help with this, what you provided so far does not reproduce. 

As said, contexts matter, if the entity is managed by different EM fork, it will end up the same way.

**Reply by milesje** [reply]
So the issue appears to be that if the value for the vehicleType is undefined at the time em.create() is called it fails attempting to perform an insert, but if I change the undefined to null (data.vehicleType = null) it works as expected....

**Reply by B4nan** [reply]
I really dont think this matters. You should realize that in your OP you are complaining about insert query that inserts more than just vehicle type PK - it also has the type and description properties - in your repro you dont specify those anywhere. They need to come from some place, and that place is where you are doing something wrong. I really do think you are passing some data in `em.create()` or similar. `undefined` and `null` should behave similarly, none of that matters for new entity, especially if you assign such relation property afterwards via `em.getReference`. 

As said multiple times, if you need more help, I need to see some [minimal but complete reproduction](https://stackoverflow.com/help/minimal-reproducible-example), not random snippets of code you think are important, something complete so its clear what you are doing.

**Reply by B4nan** [reply]
One more hint about debugging this, in case it would be about the context mismatch. You can check what EM ID is an entity managed by via `wrap(ent, true).__em.id`. But as said above, the insert query you dont like has extra data like type and description, you need to be providing that yourself somewhere. Neither `em.create()` nor `em.getReference()` are async, they can't load anything, and `em.getReference()` will indeed make the entity managed so will never produce any queries on its own (all it does it to check the identity map for already loaded entitites, if not found, creates new instance that will be ignored when flushing as it will only represent its PK).

**Reply by milesje** [reply]
So I did get things to work... I don't really like the workaround I have to do to make it do what every other ORM does automatically.
So in my case the data being received does contain the full vehicleType object (id, description, ...) and even though after calling the em.create() I was setting the entities vehicleType to a reference, because the object was there during creation it was attempting to perform an insert. So my solution was to create a temporary variable to store the vehicleType information and then set the value of data.vehicleType equal to NULL prior to the em.create() call. This allows it to work as expected. 

I'm extremely shocked at how much my SLOC has increased simply by switching from TypeORM to MikroORM. I'm talking on the magnitude of adding around 400 lines of code.

As an example here is what my code looks like for MikroORM to work how I need it to...
```
async save(customerId: string, vehicle: VehicleDto, schema: string) {
    const vehicleType = vehicle.vehicleType as VehicleTypeDto
    let entity = null
    if (vehicle.id !== undefined && vehicle.id !== null && vehicle.id !== ''){
      // get the entity from the DB as this is an update!
      entity = await this.em.findOne(Vehicle, {id: vehicle.id}, {populate: this.relations, schema})
      wrap(entity).assign(vehicle, {mergeObjects: true})
    } else {
      vehicle.vehicleType = null
      // this is an insert 
      entity = await this.em.create(Vehicle, vehicle, {schema})
      entity.customer = await this.em.getReference(Customer, customerId)
      entity.vehicleType = await this.em.getReference(VehicleType, vehicleType.id)
    }
    await this.em.persistAndFlush(entity)
    return entity
  }
```
versus TypeORM
```
async save(customerId: string, vehicle: VehicleDto, schema: string) {
  return this.vehicleRepo.save(vehicle)
}
```
I know everyone hates when one library/framework is compared to another and I'm not saying one is better than the other... It is just that every other library that I've used across multiple languages; Entity Framework, TypeORM, and hibernate to name a few all have the ability to perform simple one liners to save data... it takes multiple steps perform the same thing using this library. I'm still learning how things work differently here and will most likely create my own wrapper around mikro-orm so that my use case will look and function similar to those used by other libraries. I know this seams to be a mean rant and it is NOT my intent. Just pointing out differences to hopefully help others that stumble upon this discussion later on.

**Reply by B4nan** [reply]
I asked you several times to provide complete reproduciton, so I can finally understand what exactly are you doing wrong. So far you just keep talking, and providing very imcomplete snippets, always without the actual data you pass in. Your "workaround" is not something you should be using, but given you dont wanna provide reproduction, I have no idea how to help you. 

Btw "save" methods like this are an antipattern here, you should not flush every single insert (this literally translates to 1 query wrapped inside transaction, there is no point in doing that). You want to batch everything - which is what the ORM will do for you automatically on every flush call - but there should be just one for the whole context, not one for each entity instance. 

I can easily understand you have much more lines of code, as you are fighting the ORM, trying to use it as typeorm or similar. When I rewrote one typeorm app to mikroorm, I got the opposite result, much less LoC. Why? Because I haven't been fighting it.

So once again, if you want help, always provide complete reproductions so its clear what you are doing. code >>> words.

**Reply by milesje** [reply]
Here is a repo that reproduces my issue.  https://github.com/milesje/mikro-orm-test
I do not really understand your comment about not flushing... My use case is a web service that receives data that needs to be persisted in the database. Once that web call is complete it is expected to be available in the database for other clients and web services to be able to query from... so I would think it would need to be flushed!

The example repo does need to already have the vehicle_type and customer tables populated with data prior to running.
